ffcx/codegeneration/integral_generator.py

```diff
@@ -467,6 +476,18 @@ def generate_block_parts(
                 index = create_dof_index(table_ref, symbol)
                 B_indices.append(index)
 
+            if self.ir.part == TensorPart.diagonal and block_rank == 2:
+                assert len(A_shape) == 1
+                # If off-diagonal in mixed element, skip contribution
+                if B_indices[0].size() != B_indices[1].size():
```

Comment:
I don't think this is the right way to check for off-diagonal mixed element contributions. You can have a mixed element with same number of dofs for each subelement, just consider MixedElement([P1, P1]), which will then incorrectly add contributions from mixed terms.

I've been looking at the code for some time now, but I cannot see a solution for this yet. One has the tabledata table name, but it is unfortunately not the right thing to compare, since e.g. two names `('FE3_C0_D100_Q929', 'FE3_C0_D001_Q929')` mean access to different tables, but might just come from the same diagonal argument product. In other words, we can now compare Modified Arguments, but more work is required to store the original Argument and compare blockdata based on that.

Reply:
The best thing would be to do it at the ufl level. I started an alternative version where I tried to use ufl.extract_blocks.

however, as the argument is replaced when using it, it is also not correct. 
one could make a copy of extract_blocks that doesn’t modify the arguments, and use it prior to analysis.